src/EditorFeatures/Core/ExternalAccess/VSTypeScript/VSTypeScriptInProcLanguageClient.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor;
+using Microsoft.CodeAnalysis.Editor.Implementation.LanguageClient;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript.Api;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.LanguageServer.Client;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Utilities;
+using Newtonsoft.Json;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.VSTypeScript
+{
+    /// <summary>
+    /// Language client to handle TS LSP requests.
+    /// Allows us to move features to LSP without being blocked by TS as well
+    /// as ensures that TS LSP features use correct solution snapshots.
+    /// </summary>
+    [ContentType(ContentTypeNames.TypeScriptContentTypeName)]
+    [ContentType(ContentTypeNames.JavaScriptContentTypeName)]
+    [Export(typeof(ILanguageClient))]
+    internal class VSTypeScriptInProcLanguageClient : AbstractInProcLanguageClient
+    {
+        private readonly IVSTypeScriptCapabilitiesProvider? _typeScriptCapabilitiesProvider;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, true)]
+        public VSTypeScriptInProcLanguageClient(
+            [Import(AllowDefault = true)] IVSTypeScriptCapabilitiesProvider? typeScriptCapabilitiesProvider,
+            VSTypeScriptRequestDispatcherFactory requestDispatcherFactory,
+            IGlobalOptionService globalOptions,
+            IAsynchronousOperationListenerProvider listenerProvider,
+            LspWorkspaceRegistrationService lspWorkspaceRegistrationService,
+            DefaultCapabilitiesProvider defaultCapabilitiesProvider,
+            ILspLoggerFactory lspLoggerFactory,
+            IThreadingContext threadingContext)
+            : base(requestDispatcherFactory, globalOptions, listenerProvider, lspWorkspaceRegistrationService, lspLoggerFactory, threadingContext, diagnosticsClientName: null)
+        {
+            _typeScriptCapabilitiesProvider = typeScriptCapabilitiesProvider;
+        }
+
+        protected override ImmutableArray<string> SupportedLanguages => ImmutableArray.Create(InternalLanguageNames.TypeScript);
+
+        public override ServerCapabilities GetCapabilities(ClientCapabilities clientCapabilities)
+        {
+            var serverCapabilities = GetTypeScriptServerCapabilities(clientCapabilities);
```

Comment:
we combine any capabilities that ts provided via external access with the capabilities we wish to turn on from roslyn handlers.